**1. Write a Python program to print "Hello Python".**

In [2]:
from string import punctuation

from PIL.ImageChops import subtract
from pydantic_core.core_schema import no_info_plain_validator_function

print("Hello_World")

Hello_World


**2.Write a Python program to do arithmetical operations addition and division.**

In [4]:
#Addition
a=float(input("Enter First Number : "))
b=float(input("Enter Second Number : "))
c = a + b
print("Addition Of Two Number Is : ",c)

Addition Of Two Number Is :  132.0


In [6]:
#Division
a=float(input("Enter the dividend for division  : "))
b=float(input("Enter the divisor for division : "))
c= a/b
if b==0:
    print("Error: Division By Zero Is Not Allowed")
else:
    print(f"Division: {a} / {b} = {c}")

Division: 14.0 / 3.0 = 4.666666666666667


**3. Write a Python program to find the area of a triangle.**

In [9]:
base = float(input("Enter Base Value : "))
height = float(input("Enter Height Value "))
area = 0.5 * base * height
print(f"Base = {base} : Height = {height} = Area :{area}")

Base = 10.0 : Height = 18.0 = Area :90.0


**4. Write a Python program to swap two variables.**

In [12]:
num1=int(input("Enter First Number : "))
num2=int(input("Enter Second Number : "))
print(f"Before Swapping {num1}  and {num2}")
temp=num1
num1=num2
num2=temp
print(f"After Swapping {num1} and {num2}")

Before Swapping 10  and 20
After Swapping 20 and 10


In [14]:
a = 70
b = 20
a = a + b
b = a - b
a = a - b
print(a, b)

20 70


**5.Write a Python program to generate a random number.**

In [51]:
import random
print(f"Random Number : {random.randint(1,100)}")
print(random.random())

Random Number : 7
0.31078853634420023


**6. Write a Python program to convert kilometers to miles.**

In [54]:
km = float(input("Enter Your Value In KM : "))
miles = km * 0.621371
print(f"{km} kilometers is equal to {miles} miles")

8.0 kilometers is equal to 4.970968 miles


**7.Write a Python program to convert Celsius to Fahrenheit.**

In [55]:
celsius = float(input("Enter temperature in Celsius: "))
fahrenheit = (celsius * 9/5) + 32
print("Temperature in Fahrenheit:", fahrenheit)

Temperature in Fahrenheit: 41.0


**8.Write a Python program to display calendar.**

In [60]:
import calendar
year = int(input("Enter Year : "))
month = int(input("Enter Month : "))
cal= calendar.month(year,month)
#cal= calendar.calendar(year) # For complete year
print(cal)

    October 2026
Mo Tu We Th Fr Sa Su
          1  2  3  4
 5  6  7  8  9 10 11
12 13 14 15 16 17 18
19 20 21 22 23 24 25
26 27 28 29 30 31



**9.Write a Python program to solve quadratic equation.**

In [61]:
import math
#input coefficients
a=float(input("Enter coefficient a : "))
b=float(input("Enter coefficent b is : "))
c=float(input("Enter Coefficient c is : "))

#Input coefficients
discriminant=b**2 -4*a*c

#check if the dicrimiant is positive , negative and zero
if discriminant > 0:
    root1 = (-b + math.sqrt(discriminant)) / (2*a)
    root2 = (-b - math.sqrt(discriminant)) / (2*a)
    print(f"Root 1: {root1}")
    print(f"Root 2: {root2}")
elif discriminant == 0 :
     # One real root (repeated)
    root = -b / (2*a)
    print(f"Root: {root}")
else:
    # Complex roots
     real_part = -b / (2*a)
     imaginary_part = math.sqrt(abs(discriminant)) / (2*a)
     print(f"Root 1: {real_part} + {imaginary_part}i")
     print(f"Root 2: {real_part} - {imaginary_part}i")

Root 1: -0.16333997346592444
Root 2: -1.8366600265340758


**10.Write a Python program to swap two variables without temp variable.**

In [62]:
a = 10
b = 20
a, b = b, a
print("After swapping:")
print(a, b)

After swapping:
20 10


**11.Write a Python Program to Check if a Number is Positive, Negative or Zero.**

In [66]:
Num1=float(input("Enter A Number : "))
if Num1 == 0:
    print(f"Enter Number {Num1} Is Zero")
elif Num1 >0:
    print(f"Enter Number {Num1} Is Positive")
else:
    print(f"Enter Number {Num1} Is Negative")


Enter Number 5.0 Is Positive


**12.Write a Python Program to Check if a Number is Odd or Even.**

In [71]:
num1=float(input("Enter Number : "))
if num1 % 2 == 0:
    print("Enter Number Is Even ")
else:
    print("Enter Number Is odd ")


Enter Number Is Even 


**13.Write a Python Program to Check Leap Year.**

In [74]:
year=int(input("Enter Year : "))
# divided by 100 means century year (ending with 00)
# century year divided by 400 is leap year
if year % 400 ==0 and year % 100 ==0:
    print("Enter Year Is Leap Year")
# not divided by 100 means not a century year
# year divided by 4 is a leap year
elif year % 4 ==0 and year % 100 !=0:
    print("Enter Year Is Leap Year")
# if not divided by both 400 (century year) and 4 (not century year)
# year is not leap year
else:
    print("Enter Year Is Not Leap Year")

Enter Year Is Leap Year


**14.Write a Python Program to Check Prime Number.**

In [81]:
num1=int(input("Enter A Number : "))
if num1 <=1:
    print("Enter Number Is Not Prime Number")
else:
    for i in range(2,int(num1**0.5)+1):
        if num1 % i == 0:
            print("Not A Prime Number")
            break
    else:
        print("Prime Number")

Prime Number


**15.Write a Python Program to Print all Prime Numbers in an Interval of 1-10.**

In [85]:
for num in range(1, 11):
    if num > 1:
        for i in range(2, int(num**0.5) + 1):
            if num % i == 0:
                break
        else:
            print(num)

2
3
5
7


**16.Write a Python Program to Find the Factorial of a Number.**

In [87]:
num = int(input("Enter a number: "))
factorial = 1
if num <0:
 print("Factirial does not exist for negative numbers")
elif num == 0:
 print("Factorial of 0 is 1")
else:
 for i in range(1, num+1):
    factorial = factorial*i
 print(f'The factorial of {num} is {factorial}')

The factorial of 5 is 120


**17.Write a Python Program to Display the multiplication Table.**

In [91]:
num = int(input("Enter a number: "))
for i in range(1,11):
    print(f"{i} * {num} = {i*num}")

1 * 13 = 13
2 * 13 = 26
3 * 13 = 39
4 * 13 = 52
5 * 13 = 65
6 * 13 = 78
7 * 13 = 91
8 * 13 = 104
9 * 13 = 117
10 * 13 = 130


**18.Write a Python Program to Print the Fibonacci sequence.**

In [94]:
nterms=int(input("How many times ? "))
# first two terms
n1,n2=0,1
#check if the number of terms is valid
if nterms<=0:
    print("Please enter positive integer")
#if there is only one term , return n1
elif nterms ==1:
    print("Fibonacci sequence upto",nterms,":")
    print(n1)
else:
    print("Fibonacci Series :")
    for i in range(nterms):
        print(n1)
        nth=n1+n2
        #update values
        n1=n2
        n2=nth

Fibonacci Series :
0
1
1
2
3
5
8
13


**19.Write a Python Program to Check Armstrong Number?**


In [95]:
num = int(input("Enter a number: "))
# Calculate the number of digits in num
num_str = str(num)
num_digits = len(num_str)
# Initialize variables
sum_of_powers = 0
temp_num = num
# Calculate the sum of digits raised to the power of num_digits
while temp_num > 0:
 digit = temp_num % 10
 sum_of_powers += digit ** num_digits
 temp_num //= 10
# Check if it's an Armstrong number
if sum_of_powers == num:
 print(f"{num} is an Armstrong number.")
else:
 print(f"{num} is not an Armstrong number.")


153 is an Armstrong number.


**20.Write a Python Program to Find Armstrong Number in an Interval.**

In [96]:
# Input the interval from the user
lower = int(input("Enter the lower limit of the interval: "))
upper = int(input("Enter the upper limit of the interval: "))
for num in range(lower, upper + 1): # Iterate through the numbers i
 order = len(str(num)) # Find the number of digits in 'num'
 temp_num = num
 sum = 0
 while temp_num > 0:
  digit = temp_num % 10
  sum += digit ** order
  temp_num //= 10
  # Check if 'num' is an Armstrong number
 if num == sum:
  print(num)



1
2
3
4
5
6
7
8
9
153
370
371
407


**21.Write a Python Program to Find the Sum of Natural Numbers.**

In [1]:
limit = int(input("Enter the limit: "))
# Initialize the sum
sum = 0
# Use a for loop to calculate the sum of natural numbers
for i in range(1, limit + 1):
 sum += i
# Print the sum
print("The sum of natural numbers up to", limit, "is:", sum)

The sum of natural numbers up to 55 is: 1540


**22.Write a Python Program to Find LCM.**

In [6]:
def compute_lcm(x, y):
     if x > y: # choose the greater number
        greater = x
     else:
        greater = y
     while(True):
        if((greater % x == 0) and (greater % y == 0)):
            lcm = greater
            break
        greater += 1
     return lcm
num1 = int(input('Enter the number: '))
num2 = int(input('Enter the number: '))
print("The L.C.M. is", compute_lcm(num1, num2))


The L.C.M. is 216


**23.Write a Python Program to Find HCF.**

In [7]:
# define a function
def compute_hcf(x, y):
# choose the smaller number
 if x > y:
    smaller = y
 else:
    smaller = x
 for i in range(1, smaller+1):
    if((x % i == 0) and (y % i == 0)):
        hcf = i
 return hcf
num1 = int(input('Enter the number: '))
num2 = int(input('Enter the number: '))
print("The H.C.F. is", compute_hcf(num1, num2))


The H.C.F. is 6


**24.Write a Python Program to Convert Decimal to Binary, Octal and Hexadecimal.**

In [8]:
dec_num = int(input('Enter a decimal number: '))
print("The decimal value of", dec_num, "is:")
print(bin(dec_num), "in binary.")
print(oct(dec_num), "in octal.")
print(hex(dec_num), "in hexadecimal.")

The decimal value of 4 is:
0b100 in binary.
0o4 in octal.
0x4 in hexadecimal.


**25.Write a Python Program To Find ASCII value of a character.**

In [9]:
char = str(input("Enter the character: "))
print("The ASCII value of '" + char + "' is", ord(char))

The ASCII value of 'p' is 112


**26.Write a Python Program to Make a Simple Calculator with 4 basic mathematical
operations.**

In [3]:
num1=float(input("Enter First Number : "))
num2=float(input("Enter Second Number : "))
cal=input("Please Enter Valid Option To Perform Operations + , - , *, / ")
if cal == '+':
    add=num1+num2
    print("The Addition OF Two Number Is : ",add)
elif cal == '-':
    sub=num1-num2
    print("The Substraction OF Two Number Is : ",sub)
elif cal == '*':
    mul=num1*num2
    print("The Multiplication OF Two Number Is : ",mul)
elif cal == '/':
    div=num1/num2
    print("The Division OF Two Number Is : ",div)
else:
    print("please enter valid option ")

The Addition OF Two Number Is :  27.0


In [7]:
def add(x,y):
    return x+y
def subtract(x,y):
    return x-y
def multiply(x,y):
    return x*y
def divide(x,y):
    return x/y

print("Select Operation. ")
print("1.Add")
print("2.Subtract")
print("3.Multiply")
print("4.Divide")

while True:
    choice=input("Enter choice(1,2,3,4): ")

    if choice in ('1','2','3','4'):
            try:
                num1 = float(input("Enter First Number : "))
                num2 = float(input("Enter Second Number : "))
            except ValueError:
                print("Invalid Input. please enter a number.")
                continue

            if choice == '1':
                operation_name = "Addition"
                result = add(num1, num2)
                symbol = "+"
            elif choice == '2':
                operation_name = "Subtraction"
                result = subtract(num1, num2)
                symbol = "-"
            elif choice == '3':
                operation_name = "Multiplication"
                result = multiply(num1, num2)
                symbol = "*"
            elif choice == '4':
                operation_name = "Division"
                result = divide(num1, num2)
                symbol = "/"

    print("You selected:", operation_name)
    print(num1, symbol, num2, "=", result)

    next_calculation=input("Let's Do Next Calclulation?(yes/no):")
    if next_calculation== "no":
            break
    else:
        print("invalid input")

Select Operation. 
1.Add
2.Subtract
3.Multiply
4.Divide
You selected: Addition
20.0 + 10.0 = 30.0


**27.Write a Python Program to Display Fibonacci Sequence Using Recursion.**

In [8]:
def recur_fibo(n):
    if n <= 1:
        return n
    else:
        return recur_fibo(n-1) + recur_fibo(n-2)

nterms = int(input("Enter the number of terms (greater than 0): "))

if nterms <= 0:
    print("Please enter a positive integer")
else:
    print("Fibonacci sequence:")
    for i in range(nterms):
        print(recur_fibo(i))

Fibonacci sequence:
0
1
1
2
3


**28.Write a Python Program to Find Factorial of Number Using Recursion.**

In [9]:
def recur_factorial(n):
    if n == 1:
        return n
    else:
        return n * recur_factorial(n-1)

num = int(input("Enter the number: "))

if num < 0:
    print("Sorry, factorial does not exist for negative numbers")
elif num == 0:
    print("The factorial of 0 is 1")
else:
    print("The factorial of", num, "is", recur_factorial(num))

The factorial of 5 is 120


**29.Write a Python Program to calculate your Body Mass Index.**

In [11]:
def bodymassindex(height,weight):
    return round((weight/height**2),2)

h=float(input("Enter your heigjt in meters : "))
w=float(input("enter your weight in kg : "))

if h == 0:
    print("Height cannot be zero")
else:
    print("Welcome to the BMI calculator.")
    bmi = bodymassindex(h, w)
    print("Your BMI is:", bmi)

    if bmi <= 18.5:
        print("You are underweight.")
    elif 18.5 < bmi <= 24.9:
        print("Your weight is normal.")
    elif 25 <= bmi <= 29.9:
        print("You are overweight.")
    else:
        print("You are obese.")

Welcome to the BMI calculator.
Your BMI is: 24.44
Your weight is normal.


**30.Write a Python Program to calculate the natural logarithm of any number.**

In [12]:
import math
num = float(input("Enter a number: "))
if num <= 0:
 print("Please enter a positive number.")
else:
 # Calculate the natural logarithm (base e) of the number
 result = math.log(num)
 print(f"The natural logarithm of {num} is: {result}")

The natural logarithm of 1.78 is: 0.5766133643039938


**31.Write a Python Program for cube sum of first n natural numbers?**

In [15]:
def cube_sum_of_natural_numbers(n):
    if n <=0:
        return 0
    else:
        total= sum([i**3 for i in range(1,n+1)])
        return  total

n=int(input("Enter the value of n: "))
if n <= 0:
    print("Please enter a positive integer.")
else:
    result = cube_sum_of_natural_numbers(n)
    print(f"The cube sum of the first {n} natural numbers is: {result}")

The cube sum of the first 5 natural numbers is: 225


**32.Write a Python Program to find sum of array.**

In [17]:
arr=[1,2,3,5,6]
ans=sum(arr)
print("Sum of the array is : ",ans)

Sum of the array is :  17


In [18]:
# Function to find the sum of elements in an array
def sum_of_array(arr):
 total = 0 # Initialize a variable to store the sum

 for element in arr:
    total += element # Add each element to the total

 return total
# Example usage:
array = [1, 2, 3]
result = sum_of_array(array)
print("Sum of the array:", result)

Sum of the array: 6


**33.Write a Python Program to find largest element in an array.**

In [ ]:
def find_largest_element(arr):
    if not arr:
        return "Array Is Empty"

    largest_element=arr[0]

    for element in arr:
        if element > largest_element:
            largest_element-element
    return largest_element

my_array=[10,20,30,40,50]
result=find_largest_element(my_array)
print(f"The Largest Element In Array Is : {result}")

**34.Write a Python Program for array rotation.**

In [ ]:
def rotate_array(arr, d):
    n = len(arr)

    # Check if 'd' is valid, it should be within the range of array len
    if d < 0 or d >= n:
        return "Invalid rotation value"

    # Create a new array to store the rotated elements.
    rotated_arr = [0] * n

    # Perform the rotation.
    for i in range(n):
        rotated_arr[i] = arr[(i + d) % n]

    return rotated_arr
# Input array
arr = [1, 2, 3, 4, 5]
# Number of positions to rotate
d = 2
# Call the rotate_array function
result = rotate_array(arr, d)
# Print the rotated array
print("Original Array:", arr)
print("Rotated Array:", result)

**35.Write a Python Program to Split the array and add the first part to the end?**

In [19]:
def split_and_add(arr, k):
 if k <= 0 or k >= len(arr):
    return arr
 # Split the array into two parts
 first_part = arr[:k]
 second_part = arr[k:]
 # Add the first part to the end of the second part
 result = second_part + first_part
 return result
# Test the function
arr = [1, 2, 3, 4, 5]
k = 3
result = split_and_add(arr, k)
print("Original Array:", arr)
print("Array after splitting and adding:", result)


Original Array: [1, 2, 3, 4, 5]
Array after splitting and adding: [4, 5, 1, 2, 3]


**36.Write a Python Program to check if given array is Monotonic.**

In [20]:
def is_monotonic(arr):
    increasing = True
    decreasing = True

    for i in range(1, len(arr)):
        if arr[i] > arr[i - 1]:
            decreasing = False
        elif arr[i] < arr[i - 1]:
            increasing = False

    return increasing or decreasing


# Test the function
arr1 = [1, 2, 2, 3]  # Monotonic (non-decreasing)
arr2 = [3, 2, 1]     # Monotonic (non-increasing)
arr3 = [1, 3, 2, 4]  # Not monotonic

print("arr1 is monotonic:", is_monotonic(arr1))
print("arr2 is monotonic:", is_monotonic(arr2))
print("arr3 is monotonic:", is_monotonic(arr3))

arr1 is monotonic: True
arr2 is monotonic: True
arr3 is monotonic: False


**37.Write a Python Program to Add Two Matrices.**

In [21]:
def add_matrices(mat1, mat2):
    # Check if dimensions are same
    if len(mat1) != len(mat2) or len(mat1[0]) != len(mat2[0]):
        return "Matrices must have the same dimensions for addition"

    # Initialize result matrix
    result = []

    for i in range(len(mat1)):
        row = []
        for j in range(len(mat1[0])):
            row.append(mat1[i][j] + mat2[i][j])
        result.append(row)

    return result


# Input matrices
matrix1 = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

matrix2 = [
    [9, 8, 7],
    [6, 5, 4],
    [3, 2, 1]
]

# Call function
result_matrix = add_matrices(matrix1, matrix2)

# Print result
if isinstance(result_matrix, str):
    print(result_matrix)
else:
    print("Sum of matrices:")
    for row in result_matrix:
        print(row)

Sum of matrices:
[10, 10, 10]
[10, 10, 10]
[10, 10, 10]


**38.Write a Python Program to Multiply Two Matrices.**

In [22]:
def multiply_matrices(mat1, mat2):
    rows1 = len(mat1)
    cols1 = len(mat1[0])
    rows2 = len(mat2)
    cols2 = len(mat2[0])

    # Check if multiplication is possible
    if cols1 != rows2:
        return "Matrix multiplication is not possible"

    # Initialize result matrix with zeros
    result = [[0 for _ in range(cols2)] for _ in range(rows1)]

    # Matrix multiplication logic
    for i in range(rows1):
        for j in range(cols2):
            for k in range(cols1):
                result[i][j] += mat1[i][k] * mat2[k][j]

    return result

# Example matrices
matrix1 = [
    [1, 2, 3],
    [4, 5, 6]
]

matrix2 = [
    [7, 8],
    [9, 10],
    [11, 12]
]

# Multiply matrices
result_matrix = multiply_matrices(matrix1, matrix2)

# Print result
if isinstance(result_matrix, str):
    print(result_matrix)
else:
    print("Result of matrix multiplication:")
    for row in result_matrix:
        print(row)

Result of matrix multiplication:
[58, 64]
[139, 154]


**39.Write a Python Program to Transpose a Matrix.**

In [23]:
# Function to transpose a matrix
def transpose_matrix(matrix):
    rows, cols = len(matrix), len(matrix[0])

    # Create empty matrix (cols x rows)
    result = [[0 for _ in range(rows)] for _ in range(cols)]

    for i in range(rows):
        for j in range(cols):
            result[j][i] = matrix[i][j]

    return result

# Input matrix
matrix = [
    [1, 2, 3],
    [4, 5, 6]
]

# Transpose the matrix
transposed_matrix = transpose_matrix(matrix)

# Print result
for row in transposed_matrix:
    print(row)

[1, 4]
[2, 5]
[3, 6]


**40.Write a Python Program to Sort Words in Alphabetic Order.**

In [24]:
my_str = input("Enter a string: ")

# break string into words
words = [word.capitalize() for word in my_str.split()]

# sort the list
words.sort()

# display result
print("The sorted words are:")
for word in words:
    print(word)

The sorted words are:
Is
My
Name
Vikas


41.Write a Python Program to Remove Punctuation From a String.

In [19]:
# define punctuation
punctuations = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''

# to take input from the user
my_str = input("Enter string: ")

# remove punctuation from the string
no_punct = ""
for char in my_str:
    if char not in punctuations:
        no_punct = no_punct + char

# display the unpunctuated string
print(no_punct)

<>:2: SyntaxWarning: invalid escape sequence '\,'
<>:2: SyntaxWarning: invalid escape sequence '\,'
C:\Users\saurabh\AppData\Local\Temp\ipykernel_22736\3564093022.py:2: SyntaxWarning: invalid escape sequence '\,'
  punctuations = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''


Hi I am Vikas


**43.Write a Python program to check if the given number is a Disarium Number.**

In [27]:
def is_disarium(number):
    num_str=str(number)

    digit_sum = sum(int(i) ** (index + 1) for index, i in enumerate(num_str))

    return digit_sum==number

# Input from user
try:
    num = int(input("Enter a number: "))

    if is_disarium(num):
        print(f"{num} is a Disarium number.")
    else:
        print(f"{num} is not a Disarium number.")

except ValueError:
    print("Invalid input. Please enter a valid number.")

135 is a Disarium number.


**44.Write a Python program to print all disarium numbers between 1 to 100.**

In [28]:
def is_disarium(num):
    num_str = str(num)

    digit_sum = sum(int(i) ** (index + 1) for index, i in enumerate(num_str))

    return num == digit_sum

# Find Disarium numbers from 1 to 100
disarium_numbers = [num for num in range(1, 101) if is_disarium(num)]

print("Disarium numbers between 1 and 100:")
for num in disarium_numbers:
    print(num, end=" | ")

Disarium numbers between 1 and 100:
1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 89 | 

**45.Write a Python program to check if the given number is Happy Number.**

In [31]:
def is_happy_number(num):
    seen = set()  # store previous numbers

    while num != 1 and num not in seen:
        seen.add(num)
        num = sum(int(i) ** 2 for i in str(num))

    return num == 1

num = int(input("Enter a number: "))

if is_happy_number(num):
    print(f"{num} is a Happy Number")
else:
    print(f"{num} is not a Happy Number")

20 is not a Happy Number


**46.Write a Python program to print all happy numbers between 1 and 100.**

In [32]:
def is_happy_number(num):
    seen = set()

    while num != 1 and num not in seen:
        seen.add(num)
        num = sum(int(i) ** 2 for i in str(num))

    return num == 1

# Find happy numbers from 1 to 100
happy_numbers = []

for num in range(1, 101):
    if is_happy_number(num):
        happy_numbers.append(num)

print("Happy Numbers between 1 and 100:")
print(happy_numbers)

Happy Numbers between 1 and 100:
[1, 7, 10, 13, 19, 23, 28, 31, 32, 44, 49, 68, 70, 79, 82, 86, 91, 94, 97, 100]


**47.Write a Python program to determine whether the given number is a Harshad Number**

In [33]:
def is_harshad_number(num):
    # Calculate the sum of digits
    digit_sum = sum(int(i) for i in str(num))

    # Check divisibility
    return num % digit_sum == 0

# Input
num = int(input("Enter a number: "))

# Check
if is_harshad_number(num):
    print(f"{num} is a Harshad Number.")
else:
    print(f"{num} is not a Harshad Number.")

42 is a Harshad Number.


**48.Write a Python program to print all pronic numbers between 1 and 100.**

In [34]:
def is_pronic_number(num):
    for n in range(1, int(num**0.5) + 1):
        if n * (n + 1) == num:
            return True
    return False

print("Pronic numbers between 1 and 100 are:")

for i in range(1, 101):
    if is_pronic_number(i):
        print(i, end=" | ")

Pronic numbers between 1 and 100 are:
2 | 6 | 12 | 20 | 30 | 42 | 56 | 72 | 90 | 

**49.Sum of Elements in a List**

In [35]:
# Sample list of numbers
numbers = [10, 20, 30, 40, 50]

# Initialize a variable to store the sum
sum_of_numbers = 0

# Iterate through the list and accumulate the sum
for i in numbers:
    sum_of_numbers += i

# Print the sum
print("Sum of elements in the list:", sum_of_numbers)

Sum of elements in the list: 150


50. Write a Python program to Multiply all numbers in the list.

In [37]:
# Sample list of numbers
numbers = [10, 20, 30, 40, 50]

# Initialize a variable to store the sum
mul_of_numbers = 1

# Iterate through the list and accumulate the sum
for i in numbers:
    mul_of_numbers = mul_of_numbers*i

# Print the sum
print("Product of elements in the list:", mul_of_numbers)

Mul of elements in the list: 12000000


**51.Write a Python program to find smallest number in a list.**

In [3]:
numbers = [30,10,-45,5,20]
min=numbers[0]
for i in numbers:
    if i<min:
        min=i
print('The Smallest Number In The List Is : ',min)

The Smallest Number In The List Is :  -45


**52.Write a Python program to find largest number in a list.**

In [4]:
numbers = [30,10,-45,5,20]
max=numbers[0]
for i in numbers:
    if i>max:
        max=i
print('The Largest Number In The List Is : ',max)

The Largest Number In The List Is :  30


**53.Write a Python program to find second largest number in a list.**

In [8]:
numbers = [30,10,45,5,20]
numbers.sort(reverse=True)
print(numbers)
if len(numbers)>=2:
    second_laargest=numbers[1]
    print("The Second Largest Number In The List Is : ",second_laargest)
else:
    print("The List Does Not Contain Second Largest Number")

[45, 30, 20, 10, 5]
The Second Largest Number In The List Is :  30


**54.Write a Python program to find N largest elements from a list.**

In [9]:
def find_n_largest_elements(lst,n):
    sorted_lst=sorted(lst,reverse=True)

    largest_elements=sorted_lst[:n]

    return largest_elements

numbers=[30, 10, 45, 5, 20, 50, 15, 3, 345, 54, 67, 87, 98, 100, 34]

N = int(input("N = "))

result=find_n_largest_elements(numbers,N)

print(f"The {N} largest elements in the list are:", result)

The 4 largest elements in the list are: [345, 100, 98, 87]


**55.Write a Python program to print even numbers in a list.**

In [10]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

even_numbers=[num for num in numbers if num % 2 == 0]

print("Even Numbers In The List",even_numbers)

Even Numbers In The List [2, 4, 6, 8, 10]


**56.Write a Python program to print odd numbers in a List.**

In [11]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

odd_numbers=[num for num in numbers if num % 2 == 1]

print("Even Numbers In The List",odd_numbers)

Even Numbers In The List [1, 3, 5, 7, 9]


**57.Write a Python program to Remove empty List from List.**

In [12]:
list_of_lists=[[1,2,3],[],[4,5],[],[6,7,8],[]]
filtered_list=[i for i in list_of_lists if i]
print("List After Removing Empty List : ",filtered_list)

List After Removing Empty List :  [[1, 2, 3], [4, 5], [6, 7, 8]]


**58.Write a Python program to Cloning or Copying a list.**

In [14]:
# 1. Using Using the Slice Operator
original_list=[1,2,3,4,5]
cloned_list=original_list[:]
print(cloned_list)

[1, 2, 3, 4, 5]


In [15]:
# 2. Using the list() constructor
original_list = [1, 2, 3, 4, 5]
cloned_list = list(original_list)
print(cloned_list)

[1, 2, 3, 4, 5]


In [16]:
# 3. Using List Comprehension
original_list = [1, 2, 3, 4, 5]
cloned_list = [item for item in original_list]
print(cloned_list)

[1, 2, 3, 4, 5]


**59.Write a Python program to Count occurrences of an element in a list.**

In [20]:
def count_occurrences(l, element):
    count = l.count(element)
    return count

# Example usage:
my_list = [1, 2, 3, 4, 2, 5, 2, 3, 4, 6, 5]
element_to_count = 2

occurrences = count_occurrences(my_list, element_to_count)

print(f"The element {element_to_count} appears {occurrences} times in the list")

The element 2 appears 3 times in the list


**60.Write a Python program to find words which are greater than given length k.**

In [21]:
def find_words(words,k):
    result = []
    for i in words:
        if len(i) > k:
            result.append(i)
    return result

word_list = ["apple", "banana", "cherry", "date", "elderberry", "dragon"]
k = 5

long_words = find_words(word_list, k)

print(f"Words longer than {k} characters: {long_words}")

Words longer than 5 characters: ['banana', 'cherry', 'elderberry', 'dragon']


**61.Write a Python program for removing 𝑖th character from a string.**

In [22]:
def remove_char(input_str, i):
    # Check if i is a valid index
    if i < 0 or i >= len(input_str):
        print(f"Invalid index {i}. The string remains unchanged.")
        return input_str

    # Remove the i-th character using slicing
    result_str = input_str[:i] + input_str[i + 1:]
    return result_str

# Input string
input_str = "Hello, wWorld!"
i = 7  # Index of the character to remove

# Remove the i-th character
new_str = remove_char(input_str, i)

print(f"Original String: {input_str}")
print(f"String after removing {i}th character : {new_str}")

Original String: Hello, wWorld!
String after removing 7th character : Hello, World!


**62.Write a Python program to split and join a string.**

In [23]:
# Split a string into a list of words
input_str = "Python program to split and join a string"

word_list = input_str.split()  # By default, split on whitespace

# Join the list of words into a string
separator = " "  # specify the separator between words
output_str = separator.join(word_list)

# Print the results
print("Original String:", input_str)
print("List of split Words:", word_list)
print("Joined String:", output_str)

Original String: Python program to split and join a string
List of split Words: ['Python', 'program', 'to', 'split', 'and', 'join', 'a', 'string']
Joined String: Python program to split and join a string


**63.Write a Python program to check if a given string is binary string or not.**

In [26]:
def is_binary_str(input_str):
    # Iterate through each character in the input string
    for i in input_str:
        # Check if the character is not '0' or '1'
        if i not in '01':
            return False  # Not a binary string

    return True  # Valid binary string

# Input string to check
input_str = "110011"

# Check if the input string is a binary string
if is_binary_str(input_str):
    print(f"'{input_str}' is a binary string.")
else:
    print(f"'{input_str}' is not a binary string.")

'110011' is a binary string.


**64.Write a Python program to find uncommon words from two Strings.**

In [28]:
def uncommon_words(str1,str2):
    words1=set(str1.split())
    words2=set(str2.split())

    uncommon_words_set = words1.symmetric_difference(words2)
    uncommon_words_list = list(uncommon_words_set)
    return  uncommon_words_list

# Input two strings
string1 = "This is the first string"
string2 = "This is the second string"

uncommon = uncommon_words(string1, string2)
# Print the uncommon words
print("Uncommon words:", uncommon)

Uncommon words: ['second', 'first']


**65.Write a Python program to find all duplicate characters in string.**

In [30]:
def find_duplicates(input_str):
    # Create an empty dictionary to store character counts
    char_count = {}

    # Initialize a list to store duplicate characters
    duplicates = []

    # Count frequency of each character
    for i in input_str:
        if i in char_count:
            char_count[i] += 1
        else:
            char_count[i] = 1

    # Collect characters with count > 1
    for i, count in char_count.items():
        if count > 1:
            duplicates.append(i)

    return duplicates

# Input a string
input_string = "vikas ingawale"

# Find duplicate characters
duplicate_chars = find_duplicates(input_string)

# Print result
print("Duplicate characters:", duplicate_chars)

Duplicate characters: ['i', 'a']


**66.Write a Python Program to check if a string contains any special character.**

In [31]:
import re

def check_special_char(in_str):
    # Define a regular expression pattern to match special characters
    pattern = r'[!@#$%^&*()_+{}\[\]:;<>,.?~\\\/\'"\-=]'

    # Use re.search to find a match in the input string
    if re.search(pattern, in_str):
        return True
    else:
        return False

# Input a string
input_string = input("Enter a string: ")

# Check if the string contains any special characters
contains_special = check_special_char(input_string)

# Print the result
if contains_special:
    print("The string contains special characters.")
else:
    print("The string does not contain special characters.")

The string contains special characters.


**67.Write a Python program to Extract Unique dictionary values.**

In [33]:
# Sample dictionary
my_dict = {
 'a': 10,
 'b': 20,
 'c': 10,
 'd': 30,
 'e': 20
}
# Initialize an empty set to store unique values
uni_val = set()
# Iterate through the values of the dictionary
for i in my_dict.values():
 # Add each value to the set
 uni_val.add(i)
# Convert the set of unique values back to a list (if needed)
unique_values_list = list(uni_val)
# Print the unique values
print("Unique values in the dictionary:", unique_values_list)

Unique values in the dictionary: [10, 20, 30]


**68.Write a Python program to find the sum of all items in a dictionary.**

In [34]:
# Sample dictionary
my_dict = {
 'a': 10,
 'b': 20,
 'c': 30,
 'd': 40,
 'e': 50
}

total_sum=0

for i in my_dict.values():
    total_sum=total_sum+i

# Print the sum of all items in the dictionary
print("Sum of all items in the dictionary:", total_sum)

Sum of all items in the dictionary: 150


**69.Write a Python program to Merging two Dictionaries.**

In [35]:
# 1. Using the update() method:
dict1 = {'a': 1, 'b': 2}
dict2 = {'c': 3, 'd': 4}
dict1.update(dict2)
print("Merged Dictionary (using update()):", dict1)

Merged Dictionary (using update()): {'a': 1, 'b': 2, 'c': 3, 'd': 4}


In [ ]:
# 2. Using dictionary unpacking
dict1 = {'a': 1, 'b': 2}
dict2 = {'c': 3, 'd': 4}
# Merge dict2 into dict1 using dictionary unpacking
merged_dict = {**dict1, **dict2}
# The merged dictionary is now in merged_dict
print("Merged Dictionary (using dictionary unpacking):", merged_dict)

**70.Write a Python program to convert key-values list to flat dictionary.**

In [36]:
key_values_list = [('a', 1), ('b', 2), ('c', 3), ('d', 4)]
flat_dict={}
for key, value in key_values_list:
    flat_dict[key]=value
print("Flat Dictinary",flat_dict)

Flat Dictinary {'a': 1, 'b': 2, 'c': 3, 'd': 4}


**71.Write a Python program to insertion at the beginning in OrderedDict.**

In [2]:
from collections import OrderedDict

ordered_dict = OrderedDict([('b', 2), ('c', 3), ('d', 4)])

# Insert at beginning
ordered_dict = OrderedDict([('a', 1), *ordered_dict.items()])

print("Updated OrderedDict:", ordered_dict)

Updated OrderedDict: OrderedDict({'a': 1, 'b': 2, 'c': 3, 'd': 4})


**72.Write a Python program to check order of character in string using OrderedDict().**

In [3]:
from collections import OrderedDict

def check_order(string, reference):
    # Create OrderedDicts for both strings (keeps only unique characters in order)
    string_dict = OrderedDict.fromkeys(string)
    reference_dict = OrderedDict.fromkeys(reference)

    # Compare the order of unique characters
    return string_dict == reference_dict

# Input strings
input_string = "hello world"
reference_string = "helo wrd"

# Check if the order of characters matches
if check_order(input_string, reference_string):
    print("The order of characters in the input string matches the reference string.")
else:
    print("The order of characters in the input string does not match the reference string.")

The order of characters in the input string matches the reference string.


**73.Write a Python program to sort Python Dictionaries by Key or Value.**

In [5]:
# Sample dictionary
sample_dict = {'apple': 3, 'banana': 1, 'cherry': 2, 'date': 4}

# Sort dictionary by keys
sorted_dict_by_keys = dict(sorted(sample_dict.items()))

# Print sorted dictionary
print("Sorted by keys:")
for key, value in sorted_dict_by_keys.items():
    print(f"{key}: {value}")

Sorted by keys:
apple: 3
banana: 1
cherry: 2
date: 4


In [7]:
# Sample dictionary
sample_dict = {'apple': 3, 'banana': 1, 'cherry': 2, 'date': 4}

# Sort dictionary by values
sorted_dict_by_values = dict(sorted(sample_dict.items(), key=lambda item: item[1]))

# Print sorted dictionary
print("Sorted by values:")
for key, value in sorted_dict_by_values.items():
    print(f"{key}: {value}")

Sorted by values:
banana: 1
cherry: 2
apple: 3
date: 4


**74.**Write a program that calculates and prints the value according to the given formula:$$Q = \sqrt{\frac{2CD}{H}}$$****

In [8]:
import math

# Fixed constants
C = 50
H = 30

# Function to calculate Q using the given formula
def calculate_Q(D):
    return int(math.sqrt((2 * C * D) / H))

# Input: comma-separated D values
input_sequence = input("Enter comma-separated values of D: ")

# Split input into a list of integers
D_values = [int(D) for D in input_sequence.split(',')]

# Calculate Q for each D value
result = [calculate_Q(D) for D in D_values]

# Print results as comma-separated string
print(','.join(map(str, result)))

20


**75.Write a program which takes 2 digits, X,Y as input and generates a 2-dimensional
array. The element value in the i-th row and j-th column of the array should be i*j.**


In [11]:
# Input two digits, X and Y
X, Y = map(int, input("Enter two digits (X, Y): ").split(','))

# Initialize a 2D array filled with zeros
array = [[0 for j in range(Y)] for i in range(X)]

# Fill the array with values i * j
for i in range(X):
    for j in range(Y):
        array[i][j] = i * j

# Print the 2D array
for row in array:
    print(row)

[0, 0, 0, 0, 0]
[0, 1, 2, 3, 4]
[0, 2, 4, 6, 8]
[0, 3, 6, 9, 12]
[0, 4, 8, 12, 16]


**76.prints the words in a comma-separated sequence after sorting them alphabetically.**

**Suppose the following input is supplied to the program:
without,hello,bag,world**

**Then, the output should be:
bag,hello,without,world**

In [2]:
# Accept input from the user
input_sequence=input("Enter comma seperated sequence of words : ")
# Split the input into a list of words
words=input_sequence.split(',')
# Sort the words alphabetically
sorted_words=sorted(words)
# Join the sorted words into a comma-separated sequence
sorted_sequence=','.join(sorted_words)
# Print the sorted sequence
print("sorted words:",sorted_sequence)

sorted words: am,hello,i,vikas


**77. Write a program that accepts a sequence of whitespace separated words as input
and prints the words after removing all duplicate words and sorting them
alphanumerically.**

**Suppose the following input is supplied to the program:**

**hello world and practice makes perfect and hello world again**

**Then, the output should be:**

**again and hello makes perfect practice world**


In [3]:
#accept input from the user
input_sequence=input("Enter a sequence of whitespace-separated wordd: ")
# Split the input into words and convert it into a set to remove duplicates
words = set(input_sequence.split())
# Sort the unique words alphanumerically
sorted_words = sorted(words)
# Join the sorted words into a string with whitespace separation
result = ' '.join(sorted_words)
# Print the result
print("Result:", result)

Result: am helli hello i ingawale vikas


**79.Write a program that accepts a sentence and calculate the number of letters and digits. Suppose the following input is supplied to the program:
hello world! 123
Then, the output should be:
LETTERS 10
DIGITS 3**


In [4]:
#accept input from the user
sentence=input("Enter a sequence : ")
#Initialize counters for letters and digits
letter_count=0
digit_count=0
#Iterate through each character in the sentence
for char in sentence:
    if char.isalpha():
        letter_count+=1
    elif char.isdigit():
        digit_count+=1
# Print the results
print("LETTERS", letter_count)
print("DIGITS", digit_count)

LETTERS 13
DIGITS 4


**80.A website requires the users to input a username and password to register. Write a program to check the validity of password input by users. Following are the criteria for checking the password:**

**At least 1 letter between [a-z]**

**At least 1 letter between [A-Z]**

**At least 1 number between [0-9]**

**At least 1 character from [$#@]**

**Minimum length of password: 6**

**Maximum length of password: 12**

**Your program should accept a sequence of comma-separated passwords and check them according to the above criteria. Passwords that match the criteria are to be printed, each separated by a comma.**

In [5]:
import re

# Function to check if a password is valid
def is_valid_password(password):
    # 1. Check the length of the password (between 6 and 12 characters)
    if 6 <= len(password) <= 12:
        # 2. Check criteria: lowercase, uppercase, numbers, and special characters ($#@)
        # Added missing closing parenthesis and variable name 'password'
        if re.match(r"^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[$#@])", password):
            return True
    return False

# Accept input from the user as comma-separated passwords
raw_input = input("Enter passwords separated by commas: ")
passwords = raw_input.split(',')

# Initialize a list to store valid passwords
valid_passwords = []

# Iterate through the passwords and check their validity
for psw in passwords:
    # Strip whitespace in case user typed "pass1, pass2"
    clean_psw = psw.strip()
    if is_valid_password(clean_psw):
        valid_passwords.append(clean_psw)

# Print the valid passwords separated by commas
print("Valid Passwords:", ",".join(valid_passwords))

Valid Passwords: Vikas@6100


**81.Define a class with a generator which can iterate the numbers, which are divisible by
7, between a given range 0 and n.**


In [10]:
class DivisibleBySeven:
    def __init__(self,n):
        self.n=n
    def generate_divisible_by_seven(self):
        for num in range(self.n+1):
            if num%7==0:
                yield num

n=int(input("Enter your desired range: "))
# Fixed the truncated function name and added parentheses
obj = DivisibleBySeven(n)
divisible_by_seven_generator = obj.generate_divisible_by_seven()

# Corrected indentation for the loop
for num in divisible_by_seven_generator:
    print(num)

0
7


**82.Write a program to compute the frequency of the words from the input. The output should be displayed after sorting the keys alphanumerically. Suppose the following input is supplied to the program: > New to Python or choosing between Python 2 and Python 3? Read Python 2 or Python 3.**

Then, the output should be:

2: 2

3.: 1

3?: 1

New: 1

Python: 5

Read: 1

and: 1

between: 1

choosing: 1

or: 2

to: 1

In [12]:
input_sentence = input("Enter a sentence: ")

# Split the sentence into words
words = input_sentence.split()

# Create a dictionary to store word frequencies
word_freq = {}

# Count word frequencies
for word in words:
    # Remove punctuation (., ?) from the word
    word = word.strip('.,?')

    # Convert the word to lowercase for case-insensitive counting
    word = word.lower()

    # Update the word frequency in the dictionary
    if word in word_freq:
        word_freq[word] += 1  # Corrected indentation
    else:
        word_freq[word] = 1   # Corrected indentation

# Sort the words alphanumerically (by keys)
sorted_words = sorted(word_freq.items())

# Print the word frequencies
print("\nWord Frequencies:")
for word, frequency in sorted_words:
    print(f"{word}:{frequency}")


Word Frequencies:
guys:1
hello:2
iam:1
vikas:1


**83.Define a class Person and its two child classes: Male and Female. All classes have a
method "getGender" which can print "Male" for Male class and "Female" for Female
class.**

In [13]:
class Person:
    def getGender(self):
        return "Unknown"

class Male(Person):
    def getGender(self):
        return "Male"
class Female(Person):
    def getGender(self):
        return "Female"
person=Person()
male=Male()
female=Female()
print(person.getGender())
print(male.getGender())
print(female.getGender())

Unknown
Male
Female


**84.Please write a program to generate all sentences where subject is in ["I", "You"] and
verb is in ["Play", "Love"] and the object is in ["Hockey","Football"].**

In [14]:
subjects = ["I", "You"]
verbs = ["Play", "Love"]
objects = ["Hockey", "Football"]

sentences = []

# Triple nested loops to combine every element
for sub in subjects:
    for vrb in verbs:
        for obj in objects:
            # Create the sentence using an f-string
            sentence = f"{sub} {vrb} {obj}."
            sentences.append(sentence)

# Print the final list of sentences
for sentence in sentences:
    print(sentence)

I Play Hockey.
I Play Football.
I Love Hockey.
I Love Football.
You Play Hockey.
You Play Football.
You Love Hockey.
You Love Football.


**85.Please write a program to compress and decompress the string "hello world!hello
world!hello world!hello world!".**

In [15]:
import zlib

# 1. Define a repetitive string
text_string = "hello world!hello world!hello world!hello world!"

# 2. Compress the string
# We must .encode() the string into bytes because zlib works on binary data
compressed_data = zlib.compress(text_string.encode())

# 3. Decompress the string
# We .decode() the bytes back into a readable string (UTF-8)
decompressed_text = zlib.decompress(compressed_data).decode()

# 4. Display results
print(f"Original String: {text_string}")
print(f"Original Length: {len(text_string)} characters")
print(f"Compressed Data: {compressed_data}")
print(f"Compressed Length: {len(compressed_data)} bytes")
print(f"Decompressed String: {decompressed_text}")

Original String: hello world!hello world!hello world!hello world!
Original Length: 48 characters
Compressed Data: b'x\x9c\xcbH\xcd\xc9\xc9W(\xcf/\xcaIQ\xcc \x82\r\x00\xbd[\x11\xf5'
Compressed Length: 23 bytes
Decompressed String: hello world!hello world!hello world!hello world!


**86.Please write a binary search function which searches an item in a sorted list. The
function should return the index of element to be searched in the list.**

In [16]:
def binary_search(arr, target):
    left, right = 0, len(arr) - 1

    while left <= right:
        # Calculate the middle index
        mid = (left + right) // 2

        if arr[mid] == target:
            return mid  # Element found, return its index
        elif arr[mid] < target:
            left = mid + 1  # Target is in the right half, discard left
        else:
            right = mid - 1  # Target is in the left half, discard right

    return -1  # Element not found in the list

# Example usage:
sorted_list = [1, 2, 3, 4, 5, 6, 7, 8, 9]
target_element = 4
result = binary_search(sorted_list, target_element)

if result != -1:
    print(f"Element {target_element} found at index {result}")
else:
    print(f"Element {target_element} not found in the list")

Element 4 found at index 3


**87.Please write a program using generator to print the numbers which can be divisible
by 5 and 7 between 0 and n in comma separated form while n is input by console.
Example:
If the following n is given as input to the program:
100
Then, the output of the program should be:
0,35,70**

In [17]:
def divisible_by_5_and_7(n):
    # Loop through all numbers from 0 up to n
    for num in range(n + 1):
        # Check if number is divisible by 5 AND 7
        if num % 5 == 0 and num % 7 == 0:
            yield num

try:
    # Accept input from the user
    n = int(input("Enter a value for n: "))

    # Call the generator function
    result = divisible_by_5_and_7(n)

    # Convert results to strings and join them with commas
    print("Result:", ','.join(map(str, result)))

except ValueError:
    # Handle cases where the user types text instead of a number
    print("Invalid input. Please enter a valid integer for n.")

Result: 0,35,70


**88.Please write a program using generator to print the even numbers between 0 and n in
comma separated form while n is input by console.
Example:
If the following n is given as input to the program:
10
Then, the output of the program should be:
0,2,4,6,8,10**

In [18]:
def even_numbers(n):
    # Loop from 0 up to and including n
    for num in range(n + 1):
        # An even number has a remainder of 0 when divided by 2
        if num % 2 == 0:
            yield num

try:
    # 1. Get user input
    n = int(input("Enter a value for n: "))

    # 2. Initialize the generator
    result = even_numbers(n)

    # 3. Format and print the output
    # map(str, result) converts numbers to text so they can be joined by commas
    print("Even numbers:", ','.join(map(str, result)))

except ValueError:
    # 4. Error handling for non-integer inputs
    print("Invalid input. Please enter a valid integer for n.")

Even numbers: 0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94,96,98,100


****The** Fibonacci Sequence is computed based on the following formula:**

**f(n)=0 if n=0**

**f(n)=1 if n=1**

**f(n)=f(n-1)+f(n-2) if n>1**

**Please write a program using list comprehension to print the Fibonacci Sequence in comma separated form with a given n input by console.**

**Example:**

**If the following n is given as input to the program:**

**8**

**Then, the output of the program should be:**

**0,1,1,2,3,5,8,13**

In [19]:
def fibonacci(n):
    # The sequence always starts with 0 and 1
    if n <= 0:
        return []
    elif n == 1:
        return [0]

    sequence = [0, 1]

    # We start from index 2 because we already have indices 0 and 1
    for _ in range(2, n):
        # Add the last two numbers together and append to the list
        next_number = sequence[-1] + sequence[-2]
        sequence.append(next_number)

    return sequence

try:
    # Get the number of terms from the user
    n = int(input("Enter the number of Fibonacci terms: "))

    # Generate the sequence
    result = fibonacci(n)

    # Convert numbers to strings and join with commas for a clean display
    print("Sequence:", ','.join(map(str, result)))

except ValueError:
    print("Invalid input. Please enter a whole number (integer).")

Sequence: 0,1,1,2,3,5,8,13


**90.Assuming that we have some email addresses in the
"username@companyname.com (mailto:username@companyname.com)" format,
please write program to print the user name of a given email address. Both user
names and company names are composed of letters only.**

**Example:**

**If the following email address is given as input to the program:**

**john@google.com (mailto:john@google.com)**

**Then, the output of the program should be:**

**john**

In [20]:
def extract_username(email):
    # 1. Split the email address at '@' to separate username and domain
    # This creates a list, for example: ["user", "domain.com"]
    parts = email.split('@')

    # 2. Check if the email address has exactly one '@' symbol
    if len(parts) == 2:
        return parts[0]  # The username is the first part (index 0)
    else:
        return "Invalid email format"

try:
    # 3. Get input from the user
    email_input = input("Enter an email address: ")

    # 4. Call the function and store the result
    username = extract_username(email_input)

    # 5. Print the extracted username
    print(f"Username: {username}")

except Exception as e:
    # General error handling for unexpected issues
    print(f"An error occurred: {e}")

Username: vikas


**91.Define a class named Shape and its subclass Square. The Square class has an init
function which takes a length as argument. Both classes have an area function which
can print the area of the shape where Shape's area is 0 by default.**

In [2]:
class Shape:
    def __init__(self):
        pass
    def area(self):
        return 0
class Square(Shape):
    def __init__(self,length):
        super().__init__()
        self.length=length

    def area(self):
        return self.length ** 2

shape=Shape()
square=Square(float(input("Enter the shape of the square: ")))

# Calculate and print the areas
print("Shape's area by default:", shape.area())
print("Area of the square:", square.area())

Shape's area by default: 0
Area of the square: 25.0


**92. Write a function that stutters a word as if someone is struggling to read it. The first
two letters are repeated twice with an ellipsis ... and space after each, and then the
word is pronounced with a question mark ?.**

**Examples**

**stutter("incredible") ➞ "in... in... incredible?"**

**stutter("enthusiastic") ➞ "en... en... enthusiastic?"**

**stutter("outstanding") ➞ "ou... ou... outstanding?"**

**Hint :- Assume all input is in lower case and at least two characters long.**


In [3]:
def stutter(word):
 if len(word) < 2:
    return "Word must be at least two characters long."

 stuttered_word = f"{word[:2]}... {word[:2]}... {word}?"
 return stuttered_word

# Test cases
print(stutter("incredible"))
print(stutter("enthusiastic"))
print(stutter("outstanding"))

in... in... incredible?
en... en... enthusiastic?
ou... ou... outstanding?


**93.Create a function that takes an angle in radians and returns the corresponding angle
in degrees rounded to one decimal place.**

**Examples**

**radians_to_degrees(1) ➞ 57.3**

**radians_to_degrees(20) ➞ 1145.9**

**radians_to_degrees(50) ➞ 2864.8**


In [4]:
import math
def radians_to_degrees(radians):
    degrees=radians * (180/math.pi)
    return round(degrees,1)

# Test cases
print(radians_to_degrees(1))
print(radians_to_degrees(20))
print(radians_to_degrees(50))

57.3
1145.9
2864.8


**94.In this challenge, establish if a given integer num is a Curzon number. If 1 plus 2
elevated to num is exactly divisible by 1 plus 2 multiplied by num, then num is a
Curzon number.
Given a non-negative integer num, implement a function that returns True if num is a
Curzon number, or False otherwise.**

**Examples**

**is_curzon(5) ➞ True**

**2 ** 5 + 1 = 33**

**2 * 5 + 1 = 11**

**33 is a multiple of 11**

**is_curzon(10) ➞ False**

**2 ** 10 + 1 = 1025**

**2 * 10 + 1 = 21**

**1025 is not a multiple of 21**

**Curzon Number:
It is defined based on a specific mathematical relationship involving powers of 2. An integer
'n' is considered a Curzon number if it satisfies the following condition:
If (2^n + 1) is divisible by (2n + 1), then 'n' is a Curzon number.**

In [5]:
def is_curzon(num):
    numerator = 2 ** num + 1
    denominator = 2 * num + 1
    return numerator % denominator == 0
# Test cases
print(is_curzon(5))
print(is_curzon(10))
print(is_curzon(14))

True
False
True


**95.Given the side length x find the area of a hexagon.**

**Examples**

**area_of_hexagon(1) ➞ 2.6**

**area_of_hexagon(2) ➞ 10.4**

**area_of_hexagon(3) ➞ 23.4**

In [6]:
import math
def area_of_hexagon(x):
    area = (3 * math.sqrt(3) * x**2) / 2
    return round(area, 1)
# Example usage:
print(area_of_hexagon(1))
print(area_of_hexagon(2))
print(area_of_hexagon(3))

2.6
10.4
23.4


**96.Create a function that returns a base-2 (binary) representation of a base-10 (decimal)
string number. To convert is simple: ((2) means base-2 and (10) means base-10)
010101001(2) = 1 + 8 + 32 + 128.**

**Going from right to left, the value of the most right bit is 1, now from that every bit to
the left will be x2 the value, value of an 8 bit binary numbers are (256, 128, 64, 32, 16,
8, 4, 2, 1).**

**Examples**

**binary(1) ➞ "1"**

 **1*1 = 1**

**binary(5) ➞ "101"**

**1*1 + 1*4 = 5**

In [7]:
def binary(decimal):
     binary_str = ""
     while decimal > 0:
         remainder = decimal % 2
         binary_str = str(remainder) + binary_str
         decimal = decimal // 2
     return binary_str if binary_str else "0"

print(binary(1))
print(binary(5))
print(binary(10))

1
101
1010


**97.Create a function that takes three arguments a, b, c and returns the sum of the
numbers that are evenly divided by c from the range a, b inclusive.**

**Examples**

**evenly_divisible(1, 10, 20) ➞ 0
No number between 1 and 10 can be evenly divided by 20.**

**evenly_divisible(1, 10, 2) ➞ 30
2 + 4 + 6 + 8 + 10 = 30**

**evenly_divisible(1, 10, 3) ➞ 18
3 + 6 + 9 = 18**

In [8]:
def evenly_divisible(a,b,c):
    total=0
    for num in range(a,b+1):
         if num%c==0:
             total=total+num
    return total
print(evenly_divisible(1, 10, 20))
print(evenly_divisible(1, 10, 2))
print(evenly_divisible(1, 10, 3))

0
30
18


**98.Create a function that returns True if a given inequality expression is correct and
False otherwise.**

**Examples**

**correct_signs("3 < 7 < 11") ➞ True**

**correct_signs("13 > 44 > 33 <1") ➞ False**

**correct_signs("1 < 2 < 6 < 9 > 3") ➞ True**

In [10]:
def correct_signs(expression):
    try:
        return eval(expression)
    except:
        return False
print(correct_signs("3 < 7 < 11"))
print(correct_signs("13 > 44 > 33 < 1"))
print(correct_signs("1 < 2 < 6 < 9 > 3"))

True
False
True


**99.Create a function that replaces all the vowels in a string with a specified character.**

**Examples**

**replace_vowels("the aardvark", "#") ➞ "th# ##rdv#rk"**

**replace_vowels("minnie mouse", "?") ➞ "m?nn?? m??s?"**

**replace_vowels("shakespeare", "*") ➞ "shkspr*"**

In [12]:
def replace_vowels(string, char):
    vowels = "AEIOUaeiou"
    for vowel in vowels:
        string = string.replace(vowel, char)
    return string
print(replace_vowels("the aardvark", "#"))
print(replace_vowels("minnie mouse", "?"))
print(replace_vowels("shakespeare", "*"))

th# ##rdv#rk
m?nn?? m??s?
sh*k*sp**r*


**100.Write a function that calculates the factorial of a number recursively.**

**Examples**

**factorial(5) ➞ 120**

**factorial(3) ➞ 6**

In [13]:
def factorial(n):
    if n== 0:
        return 1
    else:
        return n * factorial(n-1)
print(factorial(5))
print(factorial(3))
print(factorial(1))
print(factorial(0))

120
6
1
1


**101.Hamming distance is the number of characters that differ between two strings.**

**To illustrate:**

**String1: "abcbba"**

**String2: "abcbda"**

**Hamming Distance: 1 - "b" vs. "d" is the only difference.**

**Create a function that computes the hamming distance between two strings.**

**Examples**

**hamming_distance("abcde", "bcdef") ➞ 5**

**hamming_distance("abcde", "abcde") ➞ 0**

In [14]:
def hamming_distance(str1, str2):
 # Check if the strings have the same length
 if len(str1) != len(str2):
    raise ValueError("Input strings must have the same length")
 # Initialize a counter to keep track of differences
 distance = 0
 # Iterate through the characters of both strings
 for i in range(len(str1)):
    if str1[i] != str2[i]:
        distance += 1 # Increment the counter for differences
 return distance

print(hamming_distance("abcde", "bcdef"))
print(hamming_distance("abcde", "abcde"))
print(hamming_distance("strong", "strung"))

5
0
1


**102.Create a function that takes a list of non-negative integers and strings and return a
new list without the strings.**

**Examples**

**filter_list([1, 2, "a", "b"]) ➞ [1, 2]**

**filter_list([1, "a", "b", 0, 15]) ➞ [1, 0, 15]**

**filter_list([1, 2, "aasf", "1", "123", 123]) ➞ [1, 2, 123]**

In [16]:
def filter_list(lst):
 # Initialize an empty list to store non-string elements
 result = []

 # Iterate through the elements in the input list
 for element in lst:
 # Check if the element is a non-negative integer (not a string)
    if isinstance(element, int) and element >= 0:
        result.append(element)

 return result

print(filter_list([1, 2, "a", "b"]))
print(filter_list([1, "a", "b", 0, 15]))

[1, 2]
[1, 0, 15]


**103.The "Reverser" takes a string as input and returns that string in reverse order, with
the opposite case.**

**Examples**

**reverse("Hello World") ➞ "DLROw OLLEh"**

**reverse("ReVeRsE") ➞ "eSrEvEr"**

**reverse("Radar") ➞ "RADAr"**

In [18]:
def reverse(input_str):
 # Reverse the string and swap the case of characters
 reversed_str = input_str[::-1].swapcase()

 return reversed_str

print(reverse("Hello World"))
print(reverse("ReVeRsE"))
print(reverse("Radar"))

DLROw OLLEh
eSrEvEr
RADAr


**104.Problem Statement: List Unpacking (Destructuring Assignment)**

**You are given a list of integers called writeyourcodehere.**

**Your task is to unpack the list into three variables using Python’s destructuring assignment:**

**first → should store the first element of the list**

**middle → should store a list of all elements except the first and last**

**last → should store the last element of the list**

In [19]:
writeyourcodehere = [1, 2, 3, 4, 5, 6]
# Unpack the list into variables
first, *middle, last = writeyourcodehere
print(first)
print(middle)
print(last)

1
[2, 3, 4, 5]
6


**105.Write a function that calculates the factorial of a number recursively.**

**Examples**

**factorial(5) ➞ 120**

**factorial(3) ➞ 6**

In [20]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return  n*factorial(n-1)

factorial(5)

120

**106.Write a function that moves all elements of one type to the end of the list.**

**Examples**

**move_to_end([1, 3, 2, 4, 4, 1], 1) ➞ [3, 2, 4, 4, 1, 1]**

**Move all the 1s to the end of the array.**

**move_to_end([7, 8, 9, 1, 2, 3, 4], 9) ➞ [7, 8, 1, 2, 3, 4, 9]**

**move_to_end(["a", "a", "a", "b"], "a") ➞ ["b", "a", "a", "a"]**

In [21]:
def move_to_end(lst, element):
     # Initialize a count for the specified element
     count = lst.count(element)

     # Remove all occurrences of the element from the list
     lst = [x for x in lst if x != element]

     # Append the element to the end of the list count times
     lst.extend([element] * count)

     return lst

print(move_to_end([1, 3, 2, 4, 4, 1], 1))
print(move_to_end([7, 8, 9, 1, 2, 3, 4], 9))

[3, 2, 4, 4, 1, 1]
[7, 8, 1, 2, 3, 4, 9]


**107.Create a function that takes a string and returns a string in which each character is
repeated once.**

**Examples**

**double_char("String") ➞ "SSttrriinngg"**

**double_char("Hello World!") ➞ "HHeelllloo WWoorrlldd!!"**

**double char("1234! ") ➞ "11223344!! "**

In [22]:
def double_char(input_str):
     doubled_str = ""

     for char in input_str:
        doubled_str += char * 2

     return doubled_str

print(double_char("String"))
print(double_char("Hello World!"))

SSttrriinngg
HHeelllloo  WWoorrlldd!!


**108.Create a function that reverses a boolean value and returns the string "boolean
expected" if another variable type is given.**

**Examples**

**reverse(True) ➞ False**

**reverse(False) ➞ True**

**reverse(0) ➞ "boolean expected"**

**reverse(None) ➞ "boolean expected"**

In [23]:
def reverse(value):
     if isinstance(value, bool):
        return not value
     else:
        return "boolean expected"
print(reverse(False))
print(reverse(True))
print(reverse(0))
print(reverse(None))

True
False
boolean expected
boolean expected


**109.Create a function that returns the thickness (in meters) of a piece of paper after
folding it n number of times. The paper starts off with a thickness of 0.5mm.**

**Examples**

**num_layers(1) ➞ "0.001m"
- Paper folded once is 1mm (equal to 0.001m)**

**num_layers(4) ➞ "0.008m"
- Paper folded 4 times is 8mm (equal to 0.008m)**

**num_layers(21) ➞ "1048.576m"
- Paper folded 21 times is 1048576mm (equal to 1048.576m)**

In [24]:
def num_layers(n):
     initial_thickness_mm = 0.5
     final_thickness_mm = initial_thickness_mm * (2 ** n)
     final_thickness_m = final_thickness_mm / 1000
     return f"{final_thickness_m:.3f}m"

print(num_layers(1))
print(num_layers(4))
print(num_layers(21))

0.001m
0.008m
1048.576m


**110.Create a function that takes a single string as argument and returns an ordered list
containing the indices of all capital letters in the string.**

**Examples**

**index_of_caps("eDaBiT") ➞ [1, 3, 5]**

**index_of_caps("eQuINoX") ➞ [1, 3, 4, 6]**

**index_of_caps("determine") ➞ []**

In [26]:
def index_of_caps(word):
     # Use list comprehension to find indices of capital letters
     return [i for i, char in enumerate(word) if char.isupper()]

print(index_of_caps("eDaBiT"))
print(index_of_caps("eQuINoX"))
print(index_of_caps("STRIKE"))
print(index_of_caps("determine"))

[1, 3, 5]
[1, 3, 4, 6]
[0, 1, 2, 3, 4, 5]
[]


**111.Using list comprehensions, create a function that finds all even numbers from 1 to
the given number.**

**Examples**

**find_even_nums(8) ➞ [2, 4, 6, 8]**

**find_even_nums(4) ➞ [2, 4]**

**find_even_nums(2) ➞ [2]**

In [27]:
def find_even_nums(num):
    return [x for x in range(1,num+1) if x % 2 == 0]
print(find_even_nums(8))
print(find_even_nums(4))
print(find_even_nums(2))

[2, 4, 6, 8]
[2, 4]
[2]


**112.Create a function that takes a list of strings and integers, and filters out the list so
that it returns a list of integers only.**

**Examples**

**filter_list([1, 2, 3, "a", "b", 4]) ➞ [1, 2, 3, 4]**

**filter_list(["A", 0, "Edabit", 1729, "Python", 1729]) ➞ [0, 1729]**

**filter_list(["Nothing", "here"]) ➞ []**

In [28]:
def filter_list(lst):
 # Use a list comprehension to filter out integers
 return [x for x in lst if isinstance(x, int)]
print(filter_list([1, 2, 3, "a", "b", 4]) )
print(filter_list(["A", 0, "Edabit", 1729, "Python", 1729]) )

[1, 2, 3, 4]
[0, 1729, 1729]


**113.Given a list of numbers, create a function which returns the list but with each
element's index in the list added to itself. This means you add 0 to the number at
index 0, add 1 to the number at index 1, etc...**

**Examples:**

**add_indexes([0, 0, 0, 0, 0]) ➞ [0, 1, 2, 3, 4]**

**add_indexes([1, 2, 3, 4, 5]) ➞ [1, 3, 5, 7, 9]**

**add_indexes([5, 4, 3, 2, 1]) ➞ [5, 5, 5, 5, 5]**

In [29]:
def add_indexes(lst):
 # Use list comprehension to add index to each element
 return [i + val for i, val in enumerate(lst)]

print(add_indexes([0, 0, 0, 0, 0]))
print(add_indexes([1, 2, 3, 4, 5]))
print(add_indexes([5, 4, 3, 2, 1]))

[0, 1, 2, 3, 4]
[1, 3, 5, 7, 9]
[5, 5, 5, 5, 5]


**114.Create a function that takes the height and radius of a cone as arguments and returns
the volume of the cone rounded to the nearest hundredth. See the resources tab for
the formula.**

**Examples**

**cone_volume(3, 2) ➞ 12.57**

**cone_volume(15, 6) ➞ 565.49**

**cone_volume(18, 0) ➞ 0**

In [31]:
import math
def cone_volume(height, radius):
     if radius == 0:
        return 0
     volume = (1/3) * math.pi * (radius**2) * height
     return round(volume, 2)

print(cone_volume(3, 2))
print(cone_volume(15, 6))
cone_volume(18, 0)

12.57
565.49


0

**115.This Triangular Number Sequence is generated from a pattern of dots that form a
triangle. The first 5 numbers of the sequence, or dots, are:**

**1, 3, 6, 10, 15**

**This means that the first triangle has just one dot, the second one has three dots, the
third one has 6 dots and so on.**

**Write a function that gives the number of dots with its corresponding triangle number
of the sequence.**

**Examples**

**triangle(1) ➞ 1**

**triangle(6) ➞ 21**

**triangle(215) ➞ 23220**

In [32]:
def triangle(n):
     if n < 1:
        return 0
     return n * (n + 1) // 2
print(triangle(1))
print(triangle(6))
print(triangle(215))

1
21
23220


**116.Create a function that takes a list of numbers between 1 and 10 (excluding one
number) and returns the missing number.**

**Examples**

**missing_num([1, 2, 3, 4, 6, 7, 8, 9, 10]) ➞ 5**

**missing_num([7, 2, 3, 6, 5, 9, 1, 4, 8]) ➞ 10**

In [2]:
def missing_num(lst):
    total_num=sum(range(1,11))
    given_sum=sum(lst)
    missing=total_num-given_sum
    return missing
print(missing_num([1, 2, 3, 4, 6, 7, 8, 9, 10]))

5


**117.Write a function that takes a list and a number as arguments. Add the number to the
end of the list, then remove the first element of the list. The function should then
return the updated list.**

**Examples**

**next_in_line([5, 6, 7, 8, 9], 1) ➞ [6, 7, 8, 9, 1]**

****next_in_line([7, 6, 3, 23, 17], 10) ➞ [6, 3, 23, 17, 10]****

**next_in_line([], 6) ➞ "No list has been selected"**

In [3]:
def next_in_line(lst,num):
    if lst:
        lst.pop(0)
        lst.append(num)
        return lst
    else:
        return "No List Has Been Selected"

print(next_in_line([5, 6, 7, 8, 9], 1) )
print(next_in_line([7, 6, 3, 23, 17], 10))

[6, 7, 8, 9, 1]
[6, 3, 23, 17, 10]


**118.Create the function that takes a list of dictionaries and returns the sum of people's
budgets.**

**Examples**

**get_budgets([
{ 'name': 'John', 'age': 21, 'budget': 23000 },
{ 'name': 'Steve', 'age': 32, 'budget': 40000 },
{ 'name': 'Martin', 'age': 16, 'budget': 2700 }
]) ➞ 65700**

In [5]:
def get_budgets(lst):
    total_budget=sum(person['budget'] for person in lst)
    return total_budget

# Test cases
budgets1 = [
 {'name': 'John', 'age': 21, 'budget': 23000},
 {'name': 'Steve', 'age': 32, 'budget': 40000},
 {'name': 'Martin', 'age': 16, 'budget': 2700}
]
get_budgets(budgets1)

65700

**119.Create a function that takes a string and returns a string with its letters in
alphabetical order.**

**Examples**

**alphabet_soup("hello") ➞ "ehllo"**

**alphabet_soup("edabit") ➞ "abdeit"**

**alphabet_soup("hacker") ➞ "acehkr"**

In [6]:
def alphabet_soup(txt):
 return ''.join(sorted(txt))
print(alphabet_soup("hello"))
print(alphabet_soup("edabit"))
print(alphabet_soup("hacker"))

ehllo
abdeit
acehkr


**120.Suppose that you invest $10,000 for 10 years at an interest rate of 6% compounded
monthly. What will be the value of your investment at the end of the 10 year period?**

**Create a function that accepts the principal p, the term in years t, the interest rate r,
and the number of compounding periods per year n. The function returns the value at
the end of term rounded to the nearest cent.**

**For the example:**

**compound_interest(10000, 10, 0.06, 12) ➞ 18193.97**

**Note that the interest rate is given as a decimal and n=12 because with monthly
compounding there are 12 periods per year. Compounding can also be done
annually, quarterly, weekly, or daily.**

**Examples**

**compound_interest(100, 1, 0.05, 1) ➞ 105.0**

**compound_interest(3500, 15, 0.1, 4) ➞ 15399.26**

In [7]:
def compound_interest(p, t, r, n):
 # Calculate the compound interest using the formula
 a = p * (1 + (r / n)) ** (n * t)
 # Round the result to the nearest cent
 return round(a, 2)
print(compound_interest(10000, 10, 0.06, 12))
print(compound_interest(100, 1, 0.05, 1))

18193.97
105.0


**121.Write a function that takes a list of elements and returns only the integers.**

**Examples**

**return_only_integer([9, 2, "space", "car", "lion", 16]) ➞ [9, 2, 16]**

**return_only_integer(["hello", 81, "basketball", 123, "fox"]) ➞ [81, 123]**

**return_only_integer(["String", True, 3.3, 1]) ➞ [1]**

In [8]:
def return_only_integer(lst):
    # Filter only integers (exclude booleans)
    return [x for x in lst if isinstance(x, int) and not isinstance(x, bool)]

print(return_only_integer([9, 2, "space", "car", "lion", 16]))
print(return_only_integer(["hello", 81, "basketball", 123, "fox"]))
return_only_integer(["String", True, 3.3, 1])

[9, 2, 16]
[81, 123]


[1]

**122.Create a function that takes three parameters where:**

**- x is the start of the range (inclusive).**

**- y is the end of the range (inclusive).**

**- n is the divisor to be checked against.**

**Return an ordered list with numbers in the range that are divisible by the third
parameter n.**

**Return an empty list if there are no numbers that are divisible by n.**

**Examples**

**list_operation(1, 10, 3) ➞ [3, 6, 9]**

**list_operation(7, 9, 2) ➞ [8]**

**list_operation(15, 20, 7) ➞ []**

In [9]:
def list_operation(x, y, n):
 # Use list comprehension to generate the list of numbers divisible
 return [num for num in range(x, y + 1) if num % n == 0]
print(list_operation(1, 10, 3))
print(list_operation(7, 9, 2))
print(list_operation(15, 20, 7))

[3, 6, 9]
[8]
[]


**123.Create a function that takes in two lists and returns True if the second list follows the
first list by one element, and False otherwise. In other words, determine if the second
list is the first list shifted to the right by 1.**

**Examples**

**simon_says([1, 2], [5, 1]) ➞ True**

**simon_says([1, 2], [5, 5]) ➞ False**

**simon_says([1, 2, 3, 4, 5], [0, 1, 2, 3, 4]) ➞ True**

**simon_says([1, 2, 3, 4, 5], [5, 5, 1, 2, 3]) ➞ False**

**Notes:**

**- Both input lists will be of the same length, and will have a minimum length of 2.**

**- The values of the 0-indexed element in the second list and the n-1th indexed
element in the first list do not matter.**


In [10]:
def simon_says(list1, list2):
 # Check if the second list is the first list shifted to the right b
 return list1[:-1] == list2[1:]
print(simon_says([1, 2], [5, 1]))
print(simon_says([1, 2], [5, 5]))
print(simon_says([1, 2, 3, 4, 5], [0, 1, 2, 3, 4]))
print(simon_says([1, 2, 3, 4, 5], [5, 5, 1, 2, 3]))

True
False
True
False


**124.A group of friends have decided to start a secret society. The name will be the first
letter of each of their names, sorted in alphabetical order. Create a function that takes
in a list of names and returns the name of the secret society.**

**Examples**

**society_name(["Adam", "Sarah", "Malcolm"]) ➞ "AMS"**

**society_name(["Harry", "Newt", "Luna", "Cho"]) ➞ "CHLN"**

**society_name(["Phoebe", "Chandler", "Rachel", "Ross", "Monica", "Joey"])**

In [12]:
def society_name(names):
 # Extract the first letter of each name, sort them, and join into a
 secret_name = ''.join(sorted([name[0] for name in names]))
 return secret_name

print(society_name(["Adam", "Sarah", "Malcolm"]))
print(society_name(["Harry", "Newt", "Luna", "Cho"]))
print(society_name(["Phoebe", "Chandler", "Rachel", "Ross", "Monica", "Joey"]))

AMS
CHLN
CJMPRR


**125.An isogram is a word that has no duplicate letters. Create a function that takes a
string and returns either True or False depending on whether or not it's an "isogram".**

**Examples**

**is_isogram("Algorism") ➞ True**

**is_isogram("PasSword") ➞ False**

**- Not case sensitive.**

**is_isogram("Consecutive") ➞ False**

**Notes
Ignore letter case (should not be case sensitive).
All test cases contain valid one word strings.**


In [13]:
def is_isogram(word):
     word = word.lower()
     # Create a set to store unique letters in the word
     unique_letters = set()

     for letter in word:
     # If the letter is already in the set, it's not an isogram
        if letter in unique_letters:
            return False
     # Otherwise, add it to the set
        unique_letters.add(letter)

     return True

print(is_isogram("Algorism"))
print(is_isogram("PasSword") )

True
False


**126.Create a function that takes a string and returns True or False, depending on whether
the characters are in order or not.**

****Examples****

**is_in_order("abc") ➞ True**

**is_in_order("edabit") ➞ False**

**is_in_order("123") ➞ True**

**is_in_order("xyzz") ➞ True**

In [14]:
def is_in_order(s):
    return s==''.join(sorted(s))
print(is_in_order("abc") )
print(is_in_order("edabit") )
print(is_in_order("123"))
print(is_in_order("xyzz"))

True
False
True
True


**127.Create a function that takes a number as an argument and returns True or False
depending on whether the number is symmetrical or not. A number is symmetrical
when it is the same as its reverse.**

**Examples**

**is_symmetrical(7227) ➞ True**

**is_symmetrical(12567) ➞ False**

**is_symmetrical(44444444) ➞ True**

**is_symmetrical(9939) ➞ False**

**is_symmetrical(1112111) ➞ True**

In [ ]:
def is_symmetrical(num):
     # Convert the number to a string
     num_str = str(num)

     # Check if the string is equal to its reverse
     return num_str == num_str[::-1]
print(is_symmetrical(7227))
print(is_symmetrical(12567))
print(is_symmetrical(44444444))

**128.Given a string of numbers separated by a comma and space, return the product of
the numbers.**

**Examples**

**multiply_nums("2, 3") ➞ 6**

**multiply_nums("1, 2, 3, 4") ➞ 24**

**multiply_nums("54, 75, 453, 0") ➞ 0**

**multiply_nums("10, -2") ➞ -20**


In [15]:
def multiply_nums(nums_str):
 # Split the input string by comma and space, then convert to intege
     nums = [int(num) for num in nums_str.split(", ")]
     # Initialize the result with 1
     result = 1

     # Multiply all the numbers together
     for num in nums:
        result *= num

     return result

print(multiply_nums("2, 3") )
print(multiply_nums("1, 2, 3, 4") )
print(multiply_nums("54, 75, 453, 0") )
print(multiply_nums("10, -2") )

6
24
0
-20


**129.Create a function that squares every digit of a number.**

**Examples**

**square_digits(9119) ➞ 811181**

**square_digits(2483) ➞ 416649**

**square_digits(3212) ➞ 9414**

**Notes
The function receives an integer and must return an integer**

In [19]:
def square_digits(n):
     # Convert the number to a string to iterate through its digits
     num_str = str(n)
     # Initialize an empty string to store the squared digits
     result_str = ""
     # Iterate through the digits
     for digit in num_str:
         # Square the digit and convert it back to an integer
         squared_digit = int(digit) ** 2
     # Append the squared digit to the result string
         result_str += str(squared_digit)
     return int(result_str)
print(square_digits(9119))
print(square_digits(2483))

811181
416649


**130.Create a function that sorts a list and removes all duplicate items from it.**

**Examples**

**setify([1, 3, 3, 5, 5]) ➞ [1, 3, 5]**

**setify([4, 4, 4, 4]) ➞ [4]**

**setify([5, 7, 8, 9, 10, 15]) ➞ [5, 7, 8, 9, 10, 15]**

In [20]:
def setify(lst):
     unique_set = set(sorted(lst))

     # Convert the set back to a list and return it
     return list(unique_set)

print(setify([1, 3, 3, 5, 5]))
print(setify([4, 4, 4, 4]) )

[1, 3, 5]
[4]


**131.Create a function that returns the mean of all digits.**

**Examples**

**mean(42) ➞ 3**

**mean(12345) ➞ 3**

**mean(666) ➞ 6**

**Notes**

**The mean of all digits is the sum of digits / how many digits there are (e.g. mean
of digits in 512 is (5+1+2)/3(number of digits) = 8/3=2).**

**The mean will always be an integer.**

In [21]:
def mean(n):
 # Convert the number to a string to iterate through its digits
 n_str = str(n)

 # Calculate the sum of digits
 digit_sum = sum(int(digit) for digit in n_str)

 # Calculate the mean by dividing the sum by the number of digits
 digit_count = len(n_str)
 digit_mean = digit_sum / digit_count

 return int(digit_mean)

print(mean(42))
print(mean(12345))
print(mean(666))

3
3
6


**132.Create a function that takes an integer and returns a list from 1 to the given number,
where:**

**1. If the number can be divided evenly by 4, amplify it by 10 (i.e. return 10 times the
number).**

**2. If the number cannot be divided evenly by 4, simply return the number.**

**Examples**

**amplify(4) ➞ [1, 2, 3, 40]**

**amplify(3) ➞ [1, 2, 3]**

**amplify(25) ➞ [1, 2, 3, 40, 5, 6, 7, 80, 9, 10, 11, 120, 13, 14, 15, 160, 17, 18, 19, 200, 21,
22, 23, 240, 25]**

**Notes**

**The given integer will always be equal to or greater than 1.
Include the number (see example above).
To perform this problem with its intended purpose, try doing it with list
comprehensions. If that's too difficult, just solve the challenge any way you can.**

In [1]:
def amplify(num):
     # Use a list comprehension to generate the list
     return [n * 10 if n % 4 == 0 else n for n in range(1, num + 1)]

print(amplify(4))
print(amplify(3))
print(amplify(25))

[1, 2, 3, 40]
[1, 2, 3]
[1, 2, 3, 40, 5, 6, 7, 80, 9, 10, 11, 120, 13, 14, 15, 160, 17, 18, 19, 200, 21, 22, 23, 240, 25]


**133.Create a function that takes a list of numbers and return the number that's unique.**

**Examples**

**unique([3, 3, 3, 7, 3, 3]) ➞ 7**

**unique([0, 0, 0.77, 0, 0]) ➞ 0.77**

**unique([0, 1, 1, 1, 1, 1, 1, 1]) ➞ 0**

**Notes**

**Test cases will always have exactly one unique number while all others are the same**

In [1]:
def unique(numbers):
     # Use a dictionary to count occurrences of each number
     count_dict = {}

     # Count occurrences of each number in the list
     for num in numbers:
         if num in count_dict:
            count_dict[num] += 1
         else:
            count_dict[num] = 1

     # Find the unique number (occurs only once)
     for num, count in count_dict.items():
        if count == 1:
            return num

print(unique([3, 3, 3, 7, 3, 3]))
print(unique([0, 0, 0.77, 0, 0]))
print(unique([0, 1, 1, 1, 1, 1, 1, 1]))

7
0.77
0


**134.Your task is to create a Circle constructor that creates a circle with a radius provided
by an argument. The circles constructed must have two getters getArea() (PIr^2) and
getPerimeter() (2PI*r) which give both respective areas and perimeter
(circumference).**

**For help with this class, I have provided you with a Rectangle constructor which you
can use as a base example.**

**Examples**

**circy = Circle(11)**

**circy.getArea()**

**Should return 380.132711084365**

**circy = Circle(4.44)**

**circy.getPerimeter()**

**Should return 27.897342763877365**

**Notes
Round results up to the nearest integer**

In [1]:
import math
class Circle:
     def __init__(self, radius):
        self.radius = radius

     def getArea(self):
     # Calculate and return the area of the circle
        return round(math.pi * self.radius**2)

     def getPerimeter(self):
     # Calculate and return the perimeter (circumference) of the cir
        return round(2 * math.pi * self.radius)

# Test cases
circy = Circle(11)
print(circy.getArea())
print(circy.getPerimeter())

380
69


**135.Create a function that takes a list of strings and return a list, sorted from shortest to
longest.**

**Examples**

**sort_by_length(["Google", "Apple", "Microsoft"]) ➞ ["Apple", "Google", "Microsoft"]**

**sort_by_length(["Leonardo", "Michelangelo", "Raphael", "Donatello"]) ➞ ["Raphael",
"Leonardo", "Donatello", "Michelangelo"]**

**sort_by_length(["Turing", "Einstein", "Jung"]) ➞ ["Jung", "Turing", "Einstein"]**

In [2]:
def sort_by_length(lst):
    return sorted(lst, key=len)

print(sort_by_length(["Google", "Apple", "Microsoft"]))
print(sort_by_length(["Leonardo", "Michelangelo", "Raphael", "Donatello"]))

['Apple', 'Google', 'Microsoft']
['Raphael', 'Leonardo', 'Donatello', 'Michelangelo']


**136.Create a function that validates whether three given integers form a Pythagorean
triplet. The sum of the squares of the two smallest integers must equal the square of
the largest number to be validated.**

**Examples**

**is_triplet(3, 4, 5) ➞ True**

**3² + 4² = 25**

**5² = 25**

**is_triplet(1, 2, 3) ➞ False**

**1² + 2² = 5**

**3² = 9**

**Notes**

**Numbers may not be given in a sorted order.**

In [3]:
def is_triplet(a, b, c):
    # Sort the numbers in ascending order
    sorted_numbers = sorted([a, b, c])

    # Check Pythagorean condition
    return sorted_numbers[0]**2 + sorted_numbers[1]**2 == sorted_numbers[2]**2


# Test cases
print(is_triplet(3, 4, 5))     # True
print(is_triplet(13, 5, 12))   # True
print(is_triplet(1, 2, 3))     # False

True
True
False


**137.Create a function that takes three integer arguments (a, b, c) and returns the amount
of integers which are of equal value.**

**Examples**

**equal(3, 4, 3) ➞ 2**

**equal(1, 1, 1) ➞ 3**

**equal(3, 4, 1) ➞ 0**

**Notes
Your function must return 0, 2 or 3.**

In [4]:
def equal(a, b, c):
     if a == b == c:
        return 3
     elif a == b or b == c or a == c:
        return 2
     else:
        return 0
print(equal(3, 4, 3))
print(equal(1, 1, 1))
print(equal(3, 4, 1))

2
3


0

**138.Write a function that converts a dictionary into a list of keys-values tuples.**

**Examples**

**dict_to_list({**

**"D": 1,**

**"B": 2**

**"C": 3**

**}) ➞ [("B", 2), ("C", 3), ("D", 1)]**

**dict_to_list({**

**"likes": 2,**

**"dislikes": 3,**

**"followers": 10**

**}) ➞ [("dislikes", 3), ("followers", 10), ("likes", 2)]**

**Notes
Return the elements in the list in alphabetical order.**

In [5]:
def dict_to_list(input_dict):
     # Sort the dictionary by keys in alphabetical order
     sorted_dict = sorted(input_dict.items())

     # Convert the sorted dictionary to a list of tuples
     result = [(key, value) for key, value in sorted_dict]

     return result

print(dict_to_list({
 "D": 1,
 "B": 2,
 "C": 3
}))

print(dict_to_list({
"likes": 2,
"dislikes": 3,
"followers": 10
}))


[('B', 2), ('C', 3), ('D', 1)]
[('dislikes', 3), ('followers', 10), ('likes', 2)]


**139.Write a function that creates a dictionary with each (key, value) pair being the (lower
case, upper case) versions of a letter, respectively.**

**Examples**

**mapping(["p", "s"]) ➞ { "p": "P", "s": "S" }**

**mapping(["a", "b", "c"]) ➞ { "a": "A", "b": "B", "c": "C" }**

**mapping(["a", "v", "y", "z"]) ➞ { "a": "A", "v": "V", "y": "Y", "z": "Z" }**

**Notes
All of the letters in the input list will always be lowercase.**

In [6]:
def mapping(letters):
 result = {}
 for letter in letters:
    result[letter] = letter.upper()
 return result
print(mapping(["p", "s"]) )
print(mapping(["a", "b", "c"]) )

{'p': 'P', 's': 'S'}
{'a': 'A', 'b': 'B', 'c': 'C'}


**140.Write a function, that replaces all vowels in a string with a specified vowel.**

**Examples**

**vow_replace("apples and bananas", "u") ➞ "upplus und bununus"**

**vow_replace("cheese casserole", "o") ➞ "chooso cossorolo"**

**vow_replace("stuffed jalapeno poppers", "e") ➞ "steffed jelepene peppers"**

**Notes**

**All words will be lowercase. Y is not considered a vowel.**

In [7]:
def vow_replace(string, vowel):
     vowels = "aeiou"
     result = ""

     for char in string:
         if char in vowels:
            result += vowel
         else:
            result += char

     return result

print(vow_replace("apples and bananas", "u"))
print(vow_replace("cheese casserole", "o"))
print(vow_replace("stuffed jalapeno poppers", "e") )

upplus und bununus
chooso cossorolo
steffed jelepene peppers


**141.Create a function that takes a string as input and capitalizes a letter if its ASCII code
is even and returns its lower case version if its ASCII code is odd.**

**Examples**

**ascii_capitalize("to be or not to be!") ➞ "To Be oR NoT To Be!"**

**ascii_capitalize("THE LITTLE MERMAID") ➞ "THe LiTTLe meRmaiD"**

**ascii_capitalize("Oh what a beautiful morning.") ➞ "oH wHaT a BeauTiFuL
moRNiNg."**

In [8]:
def ascii_capitalize(input_str):
     result = ""

     for char in input_str:
        if ord(char) % 2 == 0:
            result += char.upper()
     else:
        result += char.lower()

     return result
print(ascii_capitalize("to be or not to be!"))
print(ascii_capitalize("THE LITTLE MERMAID"))
print(ascii_capitalize("Oh what a beautiful morning.") )

T B R NT T B!
TH LTTL RDd
H HT  BTFL RNN..
